src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,571 @@
+ *     <li>The implementation should not do anything with the result of any of
+ *     the initializer, integrator, or combiner functions other than to
+ *     pass them again to the integrator, combiner, or finisher functions.</li>
+ *     <li>Once a state object is passed to the combiner or finisher function,
+ *     it is never passed to the integrator function again.</li>
+ *     <li>When the integrator function returns {@code false},
+ *     it shall be interpreted just as if there were no more elements to pass
+ *     it.</li>
+ *     <li>For parallel evaluation, the gathering implementation must manage
+ *     that the input is properly partitioned, that partitions are processed
+ *     in isolation, and combining happens only after integration is complete
+ *     for both partitions.</li>
+ *     <li>Gatherers whose combiner is {@link #defaultCombiner()} may only be
+ *     evaluated sequentially. All other combiners allow the operation to be
+ *     parallelized by initializing each partition in separation, invoking
+ *     the integrator until it returns {@code false}, and then joining each
+ *     partitions state using the combiner, and then invoking the finisher on
+ *     the joined state. Outputs and state later in the input sequence will
+ *     be discarded if processing an earlier segment short-circuits.</li>
+ *     <li>Gatherers whose finisher is {@link #defaultFinisher()} are considered
+ *     to not have an end-of-stream hook and invoking their finisher is
+ *     optional.</li>
+ * </ul>
+ *
+ * @see Stream#gather(Gatherer)
+ * @see Gatherers
+ *
+ * @param <T> the type of input elements to the gatherer operation
+ * @param <A> the potentially mutable state type of the gatherer operation
+ *            (often hidden as an implementation detail)
+ * @param <R> the type of output elements from the gatherer operation
+ * @since 22
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+public interface Gatherer<T, A, R> {
+    /**
+     * A function that produces an instance of the intermediate state used for
+     * this gathering operation.
+     *
+     * @implSpec The implementation in this interface returns
+     *           {@link #defaultInitializer()}.
+     *
+     * @return A function that produces an instance of the intermediate state
+     * used for this gathering operation
+     */
+    default Supplier<A> initializer() {
+        return defaultInitializer();
+    };
+
+    /**
+     * A function which integrates provided elements, potentially using
+     * the provided intermediate state, optionally producing output to the
+     * provided {@link Downstream}.
+     *
+     * @return a function which integrates provided elements, potentially using
+     *         the provided state, optionally producing output to the provided
+     *         Downstream
+     */
+    Integrator<A, T, R> integrator();
+
+    /**
+     * A function which accepts two intermediate states and combines them into
```

Comment:
We could use a {@return ..} here. There might be other places too where we could do that.

Reply:
Nit: I suggest using "A function that" rather than "A function which". One peculiarity with the English language is the distinction between "that" (to be used in a restrictive clause) and "which" (to be used in a non-restrictive clause). Here is a reference: https://preply.com/en/blog/when-to-use-which-or-that-the-most-common-cases

Reply:
Sounds like it is something which ought to be done wholesale rather than on a point-basis.